chain-signatures/node/src/indexer_sol.rs

```diff
@@ -0,0 +1,271 @@
+
+fn sign_request_from_event(event: SignatureRequestedEvent, tx_sig: Vec<u8>) -> anyhow::Result<IndexedSignRequest> {
+    tracing::debug!("found solana event: {:?}", event);
+    if event.deposit == 0 {
+        tracing::warn!("deposit is 0, skipping sign request");
+        return Err(anyhow::anyhow!("deposit is 0"));
+    }
+
+    if event.key_version != 0 {
+        tracing::warn!("unsupported key version: {}", event.key_version);
+        return Err(anyhow::anyhow!("unsupported key version"));
+    }
+
+    let Some(payload) = Scalar::from_bytes(event.payload) else {
+        tracing::warn!(
+            "solana `sign` did not produce payload hash correctly: {:?}",
+            event.payload,
+        );
+        return Err(anyhow::anyhow!(
+
+            "failed to convert event payload hash to scalar"
+        ));
+    };
+
+    if payload > *MAX_SECP256K1_SCALAR {
+        tracing::warn!("payload exceeds secp256k1 curve order: {payload:?}");
+        anyhow::bail!("payload exceeds secp256k1 curve order");
+    }
+
+    // Call the existing derive_epsilon_sol function with the correct parameters
+    // to match the TypeScript implementation
+    let epsilon = derive_epsilon_sol(
+        &event.sender.to_string(),
+        &event.path,
+    );
+
+    // Use transaction signature as entropy
+    let mut entropy = [0u8; 32];
+    entropy.copy_from_slice(&tx_sig[..32]);
+
+    let sign_id = SignId::new(calculate_request_id(&event));
+    tracing::info!(?sign_id, "solana signature requested");
+
+    Ok(IndexedSignRequest {
+        id: sign_id,
+        args: SignArgs {
+            entropy,
+            epsilon,
+            payload,
+            path: event.path,
+            key_version: 0,
+        },
+        chain: Chain::Solana,
+        timestamp: Instant::now(),
+    })
+}
+
+fn calculate_request_id(event: &SignatureRequestedEvent) -> [u8; 32] {
+    let mut hasher = Keccak256::new();
+    
+    // Encode the event data in a deterministic way
+    hasher.update(event.sender.to_bytes());
+    hasher.update(event.payload);
+    hasher.update(event.path.as_bytes());
+    hasher.update(&event.key_version.to_le_bytes());
+    hasher.update(&event.deposit.to_le_bytes());
+    hasher.update(&event.chain_id.to_le_bytes());
+    hasher.update(event.algo.as_bytes());
+    hasher.update(event.dest.as_bytes());
+    hasher.update(event.params.as_bytes());
+    
+    hasher.finalize().into()
+}
+
+pub async fn run(
+    sol: Option<SolConfig>,
+    sign_tx: mpsc::Sender<IndexedSignRequest>,
+    node_near_account_id: AccountId,
+) -> anyhow::Result<()> {
+    let Some(sol) = sol else {
+        tracing::warn!("solana indexer is disabled");
+        return Ok(());
+    };
+
+    tracing::info!("running solana indexer");
+    let program_id = Pubkey::from_str(&sol.program_address)?;
+    
+    let keypair = Keypair::from_base58_string(&sol.account_sk);
+
+    let cluster = Cluster::Custom(sol.rpc_url.clone(), sol.rpc_url.replace("http", "ws"));
```

Comment:
Do we need an API key?

Reply:
For production helius api key is part of the url. for local testing, shared rpc url is fine